File: packages/wasmvm/wasmlib/src/wasmtypes/scuint16.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::*;

pub const SC_UINT16_LENGTH: usize = 2;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn uint16_decode(dec: &mut WasmDecoder) -> u16 {
    uint16_from_bytes(&dec.fixed_bytes(SC_UINT16_LENGTH))
}

pub fn uint16_encode(enc: &mut WasmEncoder, value: u16) {
    enc.fixed_bytes(&uint16_to_bytes(value), SC_UINT16_LENGTH);
}

pub fn uint16_from_bytes(buf: &[u8]) -> u16 {
    let len = buf.len();
    if len == 0 {
        return 0;
    }
    if len != SC_UINT16_LENGTH {
        panic("invalid Uint16 length");
    }
    u16::from_le_bytes(buf.try_into().expect("WTF?"))
}

pub fn uint16_to_bytes(value: u16) -> Vec<u8> {
    value.to_le_bytes().to_vec()
}

pub fn uint16_from_string(value: &str) -> u16 {
    value.parse::<u16>().unwrap()
}

pub fn uint16_to_string(value: u16) -> String {
    value.to_string()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableUint16 {
    proxy: Proxy,
}

impl ScImmutableUint16 {
    pub fn new(proxy: Proxy) -> ScImmutableUint16 {
        ScImmutableUint16 { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        uint16_to_string(self.value())
    }

    pub fn value(&self) -> u16 {
        uint16_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable u16 in host container
pub struct ScMutableUint16 {
    proxy: Proxy,
}

impl ScMutableUint16 {
    pub fn new(proxy: Proxy) -> ScMutableUint16 {
        ScMutableUint16 { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: u16) {
        self.proxy.set(&uint16_to_bytes(value));
    }

    pub fn to_string(&self) -> String {
        uint16_to_string(self.value())
    }

    pub fn value(&self) -> u16 {
        uint16_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/scuint16.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

export const ScUint16Length = 2;

import {panic} from '../sandbox';
import {uintFromString, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function uint16Decode(dec: WasmDecoder): u16 {
    return dec.vluDecode(16) as u16;
}

export function uint16Encode(enc: WasmEncoder, value: u16): void {
    enc.vluEncode(value as u64);
}

export function uint16FromBytes(buf: Uint8Array): u16 {
    if (buf.length == 0) {
        return 0;
    }
    if (buf.length != ScUint16Length) {
        panic('invalid Uint16 length');
    }
    const ret: u16 = buf[1];
    return (ret << 8) | buf[0];
}

export function uint16ToBytes(value: u16): Uint8Array {
    const buf = new Uint8Array(ScUint16Length);
    buf[0] = value as u8;
    buf[1] = (value >> 8) as u8;
    return buf;
}

export function uint16FromString(value: string): u16 {
    return uintFromString(value, 16) as u16;
}

export function uint16ToString(value: u16): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableUint16 {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return uint16ToString(this.value());
    }

    value(): u16 {
        return uint16FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableUint16 extends ScImmutableUint16 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: u16): void {
        this.proxy.set(uint16ToBytes(value));
    }
}
